backend/ethereum/channel/reorg_test.go

```diff
@@ -0,0 +1,180 @@
+// limitations under the License.
+
+package channel_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+
+	"perun.network/go-perun/backend/ethereum/channel/test"
+	pkgtest "perun.network/go-perun/pkg/test"
+)
+
+// TestSimBackend_Reorg tests the reorg capability of the SimBackend.
+// It checks the assumptions that we make about how go-ethereum handles
+// chain reorganizations.
+func TestSimBackend_Reorg(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+
+	t.Run("same-order", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return txs
+		}))
+
+		// Both TX still valid.
+		s.ConfirmTx(tx1, true)
+		s.ConfirmTx(tx2, true)
+		// No events emitted.
+		s.NoMoreEvents()
+	})
+	t.Run("remove-approval", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove Approval.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{nil, nil, txs[1]}
+		}))
+		// Both TX invalid.
+		s.ConfirmTx(tx1, false)
+		s.ConfirmTx(tx2, false)
+		// All events removed.
+		s.AllowanceEvent(1, false)
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false)
+		s.NoMoreEvents()
+	})
+	t.Run("remove-transfer", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove transfer.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{txs[0]}
+		}))
+		// Wait for approval, but not transfer.
+		s.ConfirmTx(tx1, true)
+		s.ConfirmTx(tx2, false)
+		// Check that the transfer rollbacked.
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false) // allowance is 1 again
+		s.NoMoreEvents()
+	})
+	t.Run("remove-transfer-rebirth", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove transfer.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{nil, nil, txs[0]}
+		}))
+		// Wait for approval, but not transfer.
+		s.ConfirmTx(tx1, true)
+		s.ConfirmTx(tx2, false)
+		// Check that the transfer rollbacked.
+		s.AllowanceEvent(1, false) // allowance is 0
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false)
+		// Rebirth
+		s.AllowanceEvent(1, true) // allowance is 1 again
```

Comment:
I don't understand what happened to make this `1` again. Can you explain?

Reply:
`tx0` emits an `Allowance(1)` event and `tx1` an `Allowance(0)` event.  

The complexity here arises since i added two empty (`nil`) blocks.  
After the second empty block is added two the chain, the chain reorgs to the `nil, nil` chain where none of the two TX are valid; hence both `AllowanceEvents` are removed.
The third block now re-adds `tx0` and re-emitts `AllowanceEvent(1)`.